### PR TITLE
feature(cy): add dreamerV3 + MiniGrid code

### PR DESCRIPTION
**Pull request by Cloud-Pku**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/policy/mbpolicy/dreamer.py [hunk 1]
```diff
@@ -235,7 +235,8 @@ def _forward_collect(self, data: dict, world_model, envstep, reset=None, state=N
                 for i in range(len(action)):
                     action[i] *= mask[i]
 
-        data = data - 0.5
+        if type(world_model.state_size) != int and len(world_model.state_size) == 3:
```
Comment (PaParaZz1) [inline ding/policy/mbpolicy/dreamer.py:238]: isinstance

File: ding/policy/mbpolicy/dreamer.py [hunk 2]
```diff
@@ -247,11 +248,16 @@ def _forward_collect(self, data: dict, world_model, envstep, reset=None, state=N
         action = action.detach()
 
         state = (latent, action)
+        if world_model.action_type == 'discrete':
```
Comment (PaParaZz1) [inline ding/policy/mbpolicy/dreamer.py:251]: add an assertion at the first occurrence of this string comparison

File: ding/policy/mbpolicy/dreamer.py [hunk 3]
```diff
@@ -309,7 +315,8 @@ def _forward_eval(self, data: dict, world_model, reset=None, state=None) -> dict
                 for i in range(len(action)):
                     action[i] *= mask[i]
 
-        data = data - 0.5
+        if type(world_model.state_size) != int and len(world_model.state_size) == 3:
+            data = data - 0.5
```
Comment (PaParaZz1) [inline ding/policy/mbpolicy/dreamer.py:319]: add comments about this operation

File: ding/world_model/dreamer.py [hunk 4]
```diff
@@ -172,9 +192,32 @@ def train(self, env_buffer, envstep, train_iter, batch_size, batch_length):
         data = {k: torch.stack(data[k], dim=1) for k in data}  # -> {dict_key: Tensor([B, T, any_dims])}
 
         data['discount'] = data.get('discount', 1.0 - data['done'].float())
-        data['discount'] *= 0.997
+        # data['discount'] *= 0.997
```
Comment (PaParaZz1) [inline ding/world_model/dreamer.py:195]: why comment this line

File: dizoo/minigrid/config/minigrid_dreamer_config.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            discount=0.997,
+            reward_EMA=True,
+        ),
+        collect=dict(
+            n_sample=1,
+            unroll_len=1,
+            action_size=7,  # has to be specified
+            collect_dyn_sample=True,
+        ),
+        command=dict(),
```
Comment (PaParaZz1) [inline dizoo/minigrid/config/minigrid_dreamer_config.py:53]: remove unused config field

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/world_model/dreamer.py [hunk 6]
```diff
@@ -72,21 +73,26 @@ def __init__(self, cfg, env, tb_logger):
         self._cfg.norm = nn.modules.normalization.LayerNorm  # nn.LayerNorm
         self.state_size = self._cfg.state_size
         self.action_size = self._cfg.action_size
+        self.action_type = self._cfg.action_type
         self.reward_size = self._cfg.reward_size
         self.hidden_size = self._cfg.hidden_size
         self.batch_size = self._cfg.batch_size
+        if type(self.state_size) == int or len(self.state_size) == 1:
+            self.encoder = FCEncoder(self.state_size, self._cfg.encoder_hidden_size_list, activation=torch.nn.SiLU())
```
Comment (AltmanD) [inline ding/world_model/dreamer.py:81]: `state_size` here must be int. List type is not acceptable in nn.Linear() at Init of FCEncoder.

File: ding/world_model/dreamer.py [hunk 7]
```diff
@@ -113,14 +119,28 @@ def __init__(self, cfg, env, tb_logger):
...
-            self._cfg.act,
-            self._cfg.norm,
-            self.state_size,
-            self._cfg.decoder_kernels,
-        )
+
+        if type(self.state_size) == int or len(self.state_size) == 1:
+            self.heads['image'] = DenseHead(
+                feat_size,
+                (self.state_size, ),
```
Comment (AltmanD) [inline ding/world_model/dreamer.py:126]: `state_size` must be int type

File: ding/world_model/dreamer.py (hunk 4 above)
Comment (Cloud-Pku) [reply ding/world_model/dreamer.py:195]: This operation is no use. This operation only changes the model's predicted end-of-environment signal from 1 to 0.997.

File: ding/policy/mbpolicy/dreamer.py [hunk 8]
```diff
@@ -247,11 +248,17 @@ def _forward_collect(self, data: dict, world_model, envstep, reset=None, state=N
...
+        if world_model.action_type == 'discrete':
+            action = torch.where(action == 1)[1]
         output = {"action": action, "logprob": logprob, "state": state}
 
         if self._cuda:
             output = to_device(output, 'cpu')
         output = default_decollate(output)
+        if world_model.action_type == 'discrete':
+            for l in range(len(output)):
+                output[l]['action'] = output[l]['action'].squeeze(0)
```
Comment (PaParaZz1) [inline ding/policy/mbpolicy/dreamer.py:261]: can we employ squeeze operation before `default_decollate` operation? which may simplify this implementation

File: ding/world_model/dreamer.py [hunk 9]
```diff
@@ -72,21 +73,26 @@ def __init__(self, cfg, env, tb_logger):
...
+                self.state_size,
+                hidden_size_list=[32, 64, 128, 256, 4096],  # to last layer 128?
+                activation=torch.nn.SiLU(),
+                kernel_size=self._cfg.encoder_kernels,
+                layer_norm=True
+            )
+            self.embed_size = (
+                (self.state_size[1] // 2 ** (len(self._cfg.encoder_kernels))) ** 2 * self._cfg.cnn_depth *
+                2 ** (len(self._cfg.encoder_kernels) - 1)
+            )
```
Comment (PaParaZz1) [inline ding/world_model/dreamer.py:94]: add else branch for better error hints

File: ding/world_model/dreamer.py [hunk 10]
```diff
@@ -172,9 +192,31 @@ def train(self, env_buffer, envstep, train_iter, batch_size, batch_length):
...
+                    A tensor of shape [bs, num_classes, *]
+                """
+                x = x.to(torch.int64)
+                shape = (*tuple(x.shape), num_classes)
+                x = x.unsqueeze(-1)
+                res = torch.zeros(shape).to(x)
+                res = res.scatter_(-1, x, 1)
+                return res.float()
+
+            data['action'] = make_one_hot(data['action'], self.action_size)
```
Comment (PaParaZz1) [inline ding/world_model/dreamer.py:219]: why not directly use `from ding.torch_utils improt one_hot`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**